lib/url-pattern/url-match-pattern-list.js

```diff
@@ -0,0 +1,88 @@
+var _ = require('../util').lodash,
+    PropertyList = require('../collection/property-list').PropertyList,
+    Url = require('../collection/url').Url,
+    UrlMatchPattern = require('./url-match-pattern').UrlMatchPattern,
+
+    MATCH_ALL_URLS = UrlMatchPattern.MATCH_ALL_URLS,
+
+    UrlMatchPatternList;
+
+_.inherit((
+  UrlMatchPatternList = function (parent, list) {
+      UrlMatchPatternList.super_.call(this, UrlMatchPattern, parent, list);
+  }), PropertyList);
+
+_.assign(UrlMatchPatternList.prototype, /** @lends UrlMatchPatternList.prototype */ {
+    toJSON: function() {
+        if (!this.count()) {
+            return [];
+        }
+
+        return _.map(this.members, function (member) {
+            return member.toJSON();
+        });
+    },
+
+    /**
+      * Tests the url string with the match pattern list provided to see if it matches any of it.
+      * Follows the https://developer.chrome.com/extensions/match_patterns pattern for pattern validation and matching
+      *
+      * @param {String=} [urlStr] The url string for which the proxy match needs to be done.
+      * @returns {Boolean=}
+      */
+    test: function(urlStr) {
+        /*
+         * Similar to the UrlMatchPattern.test, however instead of testing
+         * MATCH_ALL_URLS and Regex conditions serially with each of the pattern,
+         * this method first searches for MATCH_ALL_URLS in all patterns
+         * and then moves on to the slower Regex based searches.
+         */
+        var url,
+            matchAllUrlsPattern,
+            matchedSpecificPattern;
+
+        matchAllUrlsPattern = this.find(function(urlMatchPattern) {
+            return urlMatchPattern.pattern === MATCH_ALL_URLS;
+        });
+
+        if (_.isObject(matchAllUrlsPattern)) {
+            return true;
+        }
+
+        try {
+            url = new Url(urlStr);
+        }
+        catch (e) {
+            return false;
+        }
+
+        matchedSpecificPattern = this.find(function(urlMatchPattern) {
+            var matchRegexObject = urlMatchPattern.createMatchPattern(urlMatchPattern.pattern);
+
+            // Empty matchRegexObject represents the match is INVALID match
+            if (_.isEmpty(matchRegexObject)) {
+                return false;
+            }
+
+            return (urlMatchPattern._testProtocol(matchRegexObject, url.protocol) &&
```

Comment:
don't leak the private functions of urlMatchPattern into urlMatchPatternList,
use 

> urlMatchPattern.test(urlString)